riptide-core/src/main/java/org/zalando/riptide/HttpResponseException.java

```diff
@@ -0,0 +1,84 @@
+package org.zalando.riptide;
+
+import com.google.common.io.ByteStreams;
+import com.google.gag.annotation.remark.Hack;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.web.client.RestClientException;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.Optional;
+
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+
+public abstract class HttpResponseException extends RestClientException {
+
+    private static final int MAX_BODY_BYTES_TO_READ = 8192;
+
+    private final int rawStatusCode;
+    private final String statusText;
+    private final HttpHeaders responseHeaders;
+    private final byte[] responseBody;
+
+    public HttpResponseException(final String message, final ClientHttpResponse response) throws IOException {
+        this(message, response.getRawStatusCode(), response.getStatusText(), response.getHeaders(),
+                extractCharset(response), readFromBody(response));
+    }
+
+    private HttpResponseException(final String message, final int rawStatusCode, final String statusText,
+            final HttpHeaders headers, final Charset charset, final byte[] responseBody) throws IOException {
+        super(format(message, responseBody, charset, rawStatusCode, statusText, headers));
+        this.rawStatusCode = rawStatusCode;
+        this.statusText = statusText;
+        this.responseHeaders = headers;
+        this.responseBody = responseBody;
+    }
+
+    private static byte[] readFromBody(final ClientHttpResponse response) throws IOException {
+        if (response.getBody() == null) {
+            return new byte[0];
+        }
+        final InputStream input = ByteStreams.limit(response.getBody(), MAX_BODY_BYTES_TO_READ);
+        final ByteArrayOutputStream output = new ByteArrayOutputStream();
+        ByteStreams.copy(input, output);
```

Comment:
Does the body have to be closed after copying?

Reply:
Fixed